src/storage/patch/N3Patcher.ts

```diff
@@ -0,0 +1,150 @@
+import { NotImplementedHttpError } from '../../util/errors/NotImplementedHttpError';
+import { readableToQuads } from '../../util/StreamUtil';
+import type { RepresentationPatcherInput } from './RepresentationPatcher';
+import { RepresentationPatcher } from './RepresentationPatcher';
+
+/**
+ * Applies an N3 Patch to a representation, or creates a new one if required.
+ * Follows all the steps from Solid, §5.3.1: https://solid.github.io/specification/protocol#n3-patch
+ */
+export class N3Patcher extends RepresentationPatcher {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly engine: ActorInitSparql;
+  private readonly generator: SparqlGenerator;
+
+  public constructor() {
+    super();
+    this.engine = newEngine();
+    this.generator = new Generator();
+  }
+
+  public async canHandle({ patch }: RepresentationPatcherInput): Promise<void> {
+    if (!isN3Patch(patch)) {
+      throw new NotImplementedHttpError('Only SPARQL update patches are supported');
+    }
+  }
+
+  public async handle(input: RepresentationPatcherInput): Promise<Representation> {
+    const patch = input.patch as N3Patch;
+
+    // No work to be done if the patch is empty
+    if (patch.deletes.length === 0 && patch.inserts.length === 0 && patch.conditions.length === 0) {
+      this.logger.debug('Empty patch, returning input.');
+      return input.representation ?? new BasicRepresentation([], input.identifier, INTERNAL_QUADS, false);
+    }
+
+    if (input.representation && input.representation.metadata.contentType !== INTERNAL_QUADS) {
+      this.logger.error('Received non-quad data. This should not happen so there is probably a configuration error.');
+      throw new InternalServerError('Quad stream was expected for patching.');
+    }
+
+    return this.patch(input);
+  }
+
+  /**
+   * Applies the given N3Patch to the representation.
+   * First the conditions are applied to find the necessary bindings,
+   * which are then applied to generate the triples that need to be deleted and inserted.
+   * After that the delete and insert operations are applied.
+   */
+  private async patch({ identifier, patch, representation }: RepresentationPatcherInput): Promise<Representation> {
+    const result = representation ? await readableToQuads(representation.data) : new Store();
+    this.logger.debug(`${result.size} quads in ${identifier.path}.`);
+
+    const { deletes, inserts } = await this.applyConditions(patch as N3Patch, identifier, result);
+
+    // Apply deletes
+    if (deletes.length > 0) {
+      // Solid, §5.3.1: "The triples resulting from ?deletions are to be removed from the RDF dataset."
+      const oldSize = result.size;
+      result.removeQuads(deletes);
+
+      // Solid, §5.3.1: "If the set of triples resulting from ?deletions is non-empty and the dataset
+      // does not contain all of these triples, the server MUST respond with a 409 status code."
+      if (oldSize - result.size !== deletes.length) {
```

Comment:
Technically, just technically, the N3 patch could contain the same triple pattern multiple times, which should be allowed. Non-trivial to detect though in case of different variable names.

Reply:
I hate that you're right because that makes this more complicated.

It actually is detectable since at this point there should no longer be any variables any more in the `deletes` array. So will have to run a check to remove duplicate triples in the array.